Windows locking crash

I'm running uv v0.1.27 on GitHub Actions on macOS x64, Linux x64, and Windows x64 runners. All jobs in my matrix succeed except for the Windows one, which reliably crashes out with an interesting OS error about a file being partially locked.

I suspect it comes from the ninja build engine aggressively invoking multiple instances of uv very closely in time; when I turn on the timestamps I see the following:
```
Tue, 02 Apr 2024 00:11:28 GMT [33/298] UV //fi.package:fi.package__editable_install(//gntools/toolchains:msvc)
Tue, 02 Apr 2024 00:11:31 GMT FAILED: bin/venv/gn_stamps/fi.package_editable_install.timestamp 
Tue, 02 Apr 2024 00:11:31 GMT C:/hostedtoolcache/windows/Python/3.11.8/x64/python.exe ../gntools/fi_run.py --stamp bin/venv/gn_stamps/fi.package_editable_install.timestamp --env VIRTUAL_ENV=bin/venv -- ../external/vendor/uv/win-x64/uv.exe -q pip install -e ../fi.package -c ../build/venv/constraints.txt
Tue, 02 Apr 2024 00:11:31 GMT error: The process cannot access the file because another process has locked a portion of the file. (os error 33)
Tue, 02 Apr 2024 00:11:31 GMT [34/298] WHEEL //fi.otherpackage:fi.otherpackage__wheel(//gntools/toolchains:msvc)
Tue, 02 Apr 2024 00:11:31 GMT [35/298] UV //fi.otherpackage:fi.otherpackage__editable_install(//gntools/toolchains:msvc)
Tue, 02 Apr 2024 00:11:31 GMT ninja: build stopped: subcommand failed.
```

The `fi_run.py` is just a wrapper that stamps a file on success, and sets optional environment variables in the process it launches (we use `VIRTUAL_ENV` to point uv at the right directory).

It looks like `fi.package` and `fi.otherpackage` are both being scheduled for concurrent execution by ninja. Each one runs `uv pip install -e path/to/package`, and the `uv` instances seem to be fighting for a lock of some kind? (The `WHEEL` command doesn't use `uv` but does use the venv to run `python -m build ...`. I don't suspect it yet...)

When I re-introduce a build lane (pool) of size 1, which forces ninja to only schedule one `uv` instance at a time, Windows succeeds. I've never seen this `os error 33` issue before, so I suspect it's new with `uv` as I introduce it into my codebase, and perhaps has something to do with locking the cache or the venv, depending on what it's doing?

I'm happy to run any experiments with custom branches if it's helpful. Thanks for uv!

--- a/crates/uv-fs/src/lib.rs
+++ b/crates/uv-fs/src/lib.rs
@@ -4,7 +4,7 @@ use std::path::{Path, PathBuf};
 use fs2::FileExt;
 use fs_err as fs;
 use tempfile::NamedTempFile;
-use tracing::{error, warn};
+use tracing::{debug, error, warn};
 
 use uv_warnings::warn_user;
 
@@ -282,18 +282,26 @@ pub struct LockedFile(fs_err::File);
 impl LockedFile {
     pub fn acquire(path: impl AsRef<Path>, resource: impl Display) -> Result<Self, std::io::Error> {
         let file = fs_err::File::create(path.as_ref())?;
+        debug!("Trying to lock if free: {}", path.as_ref().user_display());
         match file.file().try_lock_exclusive() {
             Ok(()) => Ok(Self(file)),
-            Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+            Err(err) => {
+                // Log error code and enum kind to help debugging more exotic failures
+                debug!("Try lock error, waiting for exclusive lock: {:?}", err);
                 warn_user!(
                     "Waiting to acquire lock for {} (lockfile: {})",
                     resource,
                     path.user_display(),
                 );
-                file.file().lock_exclusive()?;
+                file.file().lock_exclusive().map_err(|err| {
+                    // Not an fs_err method, we need to build our own path context
+                    std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        format!("Could not lock {}: {}", path.as_ref().user_display(), err),
+                    )
+                })?;
                 Ok(Self(file))
             }
-            Err(err) => Err(err),
         }
     }
 }

